lib/client.js

```diff
@@ -1,68 +1,37 @@
   } = {}) {
-    // TODO: allow backoff options to be passed in
-    const backoff = new Backoff({
-      min: config.redis.retry.min,
-      max: config.redis.retry.max,
-      jitter: config.redis.retry.jitter,
-      factor: config.redis.retry.factor
-    });
-    logger.verbose(`initializing client: ${host}:${port}`);
-    this.client = redis.createClient({
-      host,
-      port,
-      prefix,
-      retry_strategy: options => {
-        if(options.error && options.error.code === 'ECONNREFUSED') {
-          // end reconnecting on a specific error and flush all commands with an
-          // individual error
-          logger.error('connection refused');
-        }
-        if(options.total_retry_time > config.redis.retry.maxTotal) {
-          // end reconnecting after a specific timeout and flush all commands
-          // with a individual error
-          logger.error('exceeded max connection retry time', {
-            totalRetryTime: options.total_retry_time,
-            timesConnected: options.times_connected
-          });
-          // FIXME: improve error handling
-          return new Error('redis: exceeded max connection retry time');
-        }
-        if(options.times_connected > config.redis.retry.maxTimesConnected) {
-          // end reconnecting
-          logger.error('exceeded max connection times', {
-            totalRetryTime: options.total_retry_time,
-            timesConnected: options.times_connected
-          });
-          // FIXME: improve error handling
-          return new Error('redis: exceed max connection times');
+    logger.debug(`initializing client at: ${host}:${port}`);
+    this.client = createClient({
+      socket: {host, port, reconnectStrategy: attempt => {
+        if(attempt > config.redis.retryOptions.retries) {
+          return new Error('Exceeded max connection attempts.');
         }
-        // reconnect delay with backoff
-        if(options.attempt === 1) {
-          backoff.reset();
-        }
-        const delay = backoff.duration();
-        logger.info(`will reconnect in ${delay}ms`);
-        return delay;
+        // Calculates a timeout based on current attempt and
+        // retry options from this modules config.
+        return retry.createTimeout(attempt, config.redis.retryOptions);
       }
-    });
+      }});
+    this.client.on('ready', () => logger.debug(`Connected to redis.`));
+    this.client.on('error', error => logger.error(`Redis error: ${error}.`));
+    this.client.on('end', () => logger.debug('Connection to redis closed.'));
+  }
 
-    // FIXME: improve error handling
-    this.client.on('error', error => logger.error(`redis error: ${error}`));
-    this.client.on('end', () => logger.verbose('connection closed'));
-    this.client.on('warning', msg => logger.warning(`redis warning: ${msg}`));
+  async connect() {
+    try {
+      await this.client.connect();
+    } catch(error) {
+      logger.error('Unable to connect to redis.', {error});
```

Comment:
This seems like a problem -- if there's an error when connecting it is never surfaced outside of this call. This should rethrow the error:

```suggestion
      logger.error('Unable to connect to redis.', {error});
      throw error;
```

Reply:
resolved here 8dc411e30f97b16e9ab4cd971ee62b1922422e1b